refactor: unify container presence checks

l0rinc opened the pull request:
### Summary
Instead of counting occurrences in sets and maps, the C++20 `::contains` method expresses the intent unambiguously and can return early on first encounter.

### Context
Applied clang‑tidy's [readability‑container‑contains](https://clang.llvm.org/extra/clang-tidy/checks/readability/container-contains.html) check, though many cases required manual changes since tidy couldn't fix them automatically.

### Changes
The changes made here were:

| From                   | To               |
|------------------------|------------------|
| `m.find(k) == m.end()` | `!m.contains(k)` |
| `m.find(k) != m.end()` | `m.contains(k)`  |
| `m.count(k)`           | `m.contains(k)`  |
| `!m.count(k)`          | `!m.contains(k)` |
| `m.count(k) == 0`      | `!m.contains(k)` |
| `m.count(k) != 1`      | `!m.contains(k)` |
| `m.count(k) == 1`      | `m.contains(k)`  |
| `m.count(k) < 1`       | `!m.contains(k)`  |
| `m.count(k) > 0`       | `m.contains(k)`  |
| `m.count(k) != 0`      | `m.contains(k)`  |

> Note that `== 1`/`!= 1`/`< 1` only apply to simple [maps](https://en.cppreference.com/w/cpp/container/map/contains)/[sets](https://en.cppreference.com/w/cpp/container/set/contains) and had to be changed manually.

There are many other cases that could be changed, but we've reverted most of those to reduce conflict with other open PRs.

-----

<details>
<summary>clang-tidy command on Mac</summary>

```bash
rm -rfd build && \
cmake -B build \
  -DCMAKE_C_COMPILER="$(brew --prefix llvm)/bin/clang" \
  -DCMAKE_CXX_COMPILER="$(brew --prefix llvm)/bin/clang++" \
  -DCMAKE_OSX_SYSROOT="$(xcrun --show-sdk-path)" \
  -DCMAKE_C_FLAGS="-target arm64-apple-macos11" \
  -DCMAKE_CXX_FLAGS="-target arm64-apple-macos11" \
  -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DBUILD_BENCH=ON -DBUILD_FUZZ_BINARY=ON -DBUILD_FOR_FUZZING=ON

 "$(brew --prefix llvm)/bin/run-clang-tidy" -quiet -p build -j$(nproc) -checks='-*,readability-container-contains' | grep -v 'clang-tidy'
```

</details>

pablomartin4btc reviewed (commented):
Concept ACK

This refactor replaces explicit `.count(x)` checks with `.contains(x)`, which only verifies that at least one instance exists (or none when `== 0`). Please double-check whether `.contains()` preserves the original behavior in some contexts where exactness matters ( !=1, ==1, etc). Also where the logic enforces that exactly one item is/ is not present — useful even in containers that don't allow duplicates (like `std::map` or `std::set`) because it clearly expresses the expected logic — replacing it with `.contains()` weakens the check and may mask issues where the assumption of uniqueness is violated.

l0rinc committed:
refactor: unify container presence checks - find

The changes made here were:

| From                   | To               |
|------------------------|------------------|
| `m.find(k) == m.end()` | `!m.contains(k)` |
| `m.find(k) != m.end()` | `m.contains(k)`  |

l0rinc committed:
refactor: unify container presence checks - non-trivial counts

The changes made here were:

| From              | To               |
|-------------------|------------------|
| `m.count(k) == 1` | `m.contains(k)`  |
| `m.count(k) != 1` | `!m.contains(k)` |
| `m.count(k) < 1`  | `!m.contains(k)` |

* `mapInfo` is instance of `std::unordered_map` and can only contain 0 or 1 value for a given key;
* similarly, `g_enabled_filter_types` and `setClientRules` are both `std::set` instances;
* lastly, while `mapTxSpends` is `std::unordered_multimap` that could potentially hold multiple values, having a size less than 1 means that the value is missing.

`QMap<WalletModel*, WalletView*> mapWalletViews` values were also migrated manually.

Co-authored-by: pablomartin4btc <pablomartin4btc@gmail.com>
Co-authored-by: fanquake <fanquake@gmail.com>

l0rinc reviewed (commented):
Thanks, split up the PR into 3 commits to simplify assessing the risks - hope this clarifies it, see the commit messages for further details.

l0rinc commented on src/node/miner.cpp:
```diff
@@ -337,7 +337,7 @@ void BlockAssembler::addPackageTxs(int& nPackagesSelected, int& nDescendantsUpda
             auto it = mempool.mapTx.project<0>(mi);
             assert(it != mempool.mapTx.end());
-            if (mapModifiedTx.count(it) || inBlock.count(it->GetSharedTx()->GetHash()) || failedTx.count(it->GetSharedTx()->GetHash())) {
+            if (mapModifiedTx.count(it) || inBlock.contains(it->GetSharedTx()->GetHash()) || failedTx.contains(it->GetSharedTx()->GetHash())) {
```
I had these at one point, but decided against modifying multi-containers - but this one seems fine, thanks, added.

janb84 commented:
Nit; Missed?: 
[bitcoin-cli.cpp L11117](https://github.com/l0rinc/bitcoin/blob/71911a5912c0778b10308b4e28ff183921f4a275/src/bitcoin-cli.cpp#L1117)
``` C++
if (proxy_networks.find(proxy) == proxy_networks.end()) ordered_proxies.push_back(proxy);
``` 
=>

``` C++
if (!proxy_networks.contains(proxy)) ordered_proxies.push_back(proxy);
```

janb84 reviewed (commented):
ACK 53461dc5cedf5f7080c09246ef76ba8a6c38a103

This PR refactors the code to use the more modern contains() method. In my opinion this PR increased the readability of the code and removes the ambiguity of the intention of the count() methods used. With this change, the intent to enforce that exactly one item is/ is not present  or  just a presence check will be more obvious. (count() vs contains()).  

The argument NOT to change the code because of the risk losing the original behaviour is an indication to me that the refactor is needed to remove the ambiguity and more clearly communicate the intent of the code. 

(thanks for including my suggestion/nit)

janb84 reviewed (commented):
re ACK 05153600af10c322e40d107b8cf0c8e69b8b0aeb

changes since last ACK:
- clang-tidy related changes 

<details>
New result of clang-tidy : 

``` shell
$ ( cd ./src/ && run-clang-tidy -p ../build -j $(nproc) ) | grep walletframe.cpp
<< no warnings>> 
```

</details>

l0rinc commented:
Yes, applies to any kind of map, but I had to find and change those manually - clang-tidy doesn't detect or fix those.

optout21 reviewed (commented):
ACK 05153600af10c322e40d107b8cf0c8e69b8b0aeb

- The changes increase code readability, as 'contains' expresses the code logic / intent more specifically
- It also results in higher performance, due to potential early exit. The improvement is probably negligible though.
- Changes are localized (each to a single line), local impact only

(I've reviewed the proposed changed, but didn't look for other potential changes)

ajtowns commented:
> Not sure what to do about the conflicts, but to enable the linter, I had to fix the existing problems first. Maybe there is a way in clang-tidy to only check new code,

Perhaps an approach like this would work (for this and other cleanups)?

1. Run `git diff ${base}..HEAD --unified=0` and use the `^[+][+][+]` lines to track the files touched, and the "+n,k" values in the `^@@` lines to identify the touched lines (ie, each line in `range(n,n+k)`
2. Run clang-tidy
3. Check for any clang-tidy errors that hit lines found in step 1, and error out if there are any hits

l0rinc committed:
refactor: unify container presence checks - trivial counts

The changes made here were:

| From              | To               |
|-------------------|------------------|
| `m.count(k)`      | `m.contains(k)`  |
| `!m.count(k)`     | `!m.contains(k)` |
| `m.count(k) == 0` | `!m.contains(k)` |
| `m.count(k) != 0` | `m.contains(k)`  |
| `m.count(k) > 0`  | `m.contains(k)`  |

The commit contains the trivial, mechanical refactors where it doesn't matter if the container can have multiple elements or not

Co-authored-by: Jan B <608446+janb84@users.noreply.github.com>

l0rinc commented:
Removed all conflicts with other PR that I found and opened a new PR at https://github.com/bitcoin/bitcoin/pull/33192 to given @DrahtBot a chance to recalculate the conflicts section.
cc: @ajtowns, @instagibbs
